src/tasks/git_move_upstream.rs

```diff
@@ -0,0 +1,201 @@
+            if urls.iter().any(|url| url == &expected_url) {
+                git_remote_set_url(&repo.get_path(), remote, &new_url).await?;
+            }
+        }
+
+        Ok(())
+    }
+
+    #[tracing::instrument(name = "task:git_rename(scratchpad)", err, skip(self, _core))]
+    async fn apply_scratchpad(
+        &self,
+        _core: &Core,
+        _scratch: &crate::core::Scratchpad,
+    ) -> Result<(), crate::core::Error> {
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::core::Core;
+    use crate::git::{git_remote_add, git_remote_rename};
+    use crate::tasks::{GitInit, GitRemote};
+    use tempfile::tempdir;
+
+    #[tokio::test]
+    async fn move_upstream_errors_if_repo_does_not_exist() {
+        let temp = tempdir().unwrap();
+        let core = Core::builder()
+            .with_config_for_dev_directory(temp.path())
+            .with_null_console()
+            .build();
+
+        let repo = core
+            .resolver()
+            .get_best_repo("gh:git-fixtures/basic")
+            .unwrap();
+
+        let new_repo = core
+            .resolver()
+            .get_best_repo("gh:git-fixtures/renamed")
+            .unwrap();
+
+        let task = GitMoveUpstream {
+            new_repo: new_repo.clone(),
+        };
+
+        let result = task.apply_repo(&core, &repo).await;
+        assert!(result.is_err());
+        let err_msg = format!("{result:?}");
+        assert!(
+            err_msg.contains("does not exist"),
+            "Unexpected error message: {err_msg}"
+        );
+    }
+
+    #[tokio::test]
+    async fn move_upstream_updates_singular_remote() {
+        let temp = tempdir().unwrap();
+        let core = Core::builder()
+            .with_config_for_dev_directory(temp.path())
+            .with_null_console()
+            .build();
+
+        let repo = core
+            .resolver()
+            .get_best_repo("gh:git-fixtures/basic")
+            .unwrap();
+
+        let new_repo = core
+            .resolver()
+            .get_best_repo("gh:git-fixtures/renamed")
+            .unwrap();
+
+        sequence!(GitInit {}, GitRemote { name: "origin" })
+            .apply_repo(&core, &repo)
+            .await
+            .unwrap();
+
+        let task = GitMoveUpstream {
+            new_repo: new_repo.clone(),
+        };
+
+        let result = task.apply_repo(&core, &repo).await;
+
+        assert!(result.is_ok());
+
+        let remotes = crate::git::git_remote_list(&repo.path).await.unwrap();
+        for remote in remotes {
+            let urls = crate::git::git_remote_get_url(&repo.path, &remote)
+                .await
+                .unwrap();
+            assert!(
+                urls.iter().any(|u| u.contains("git-fixtures/renamed")),
+                "Remote was not updated to point to new repo"
+            );
+        }
+    }
+
+    #[tokio::test]
+    async fn move_upstream_updates_forked_remote() {
```

Comment:
Another test which needs `#[cfg_attr(feature = "pure-tests", ignore)]`

Reply:
refactored to be a pure test